### PR TITLE
GH-44058: [CI][Integration] Group logs on GitHub Actions

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

The current log is difficult to read.

### What changes are included in this PR?

Group logs.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.
* GitHub Issue: #44058

### CODE REVIEW

File: ci/scripts/integration_arrow.sh [hunk 1]
```diff
@@ -30,15 +30,23 @@ gold_dir=$arrow_dir/testing/data/arrow-ipc-stream/integration
...
 : ${ARROW_INTEGRATION_JS:=ON}
 
+echo "::group::Integration: Prepare: Archery"
+set -x
 pip install -e $arrow_dir/dev/archery[integration]
+set +x
+echo "::endgroup::"
 
+echo "::group::Integration: Prepare: Dependencies"
+set -x
```
Comment (pitrou) [inline ci/scripts/integration_arrow.sh:40]: I don't know if it's possible to have some kind of shell script function to factor these out?

File: dev/archery/archery/utils/logger.py [hunk 2]
```diff
@@ -27,3 +29,21 @@ def __init__(self, quiet=False):
...
+
+@contextmanager
+def group(name):
+    """
+    Group outputs in the given with block.
+
+    This does nothing in non GitHub Actions environment for now.
+    """
+    if in_github_actions:
+        print(f"::group::{name}")
```
Comment (kou) [inline dev/archery/archery/utils/logger.py:44]: Hmm. This isn't outputted immediately...
Do we need to `flush` or something...?

File: ci/scripts/integration_arrow.sh (hunk 1 above)
Comment (kou) [reply ci/scripts/integration_arrow.sh]: Hmm. I couldn't create enough convenient functions as I imagined...

File: dev/archery/archery/utils/logger.py (hunk 2 above)
Comment (kou) [reply dev/archery/archery/utils/logger.py]: Wow! I missed it.

Can we move the `log` and `_Printer` to `archery.utils.printer` or something from `archery.integration.util`?
I want to use it in `archery docker` too.

Comment (pitrou) [reply dev/archery/archery/utils/logger.py]: Then I would suggest something like:
```python

@contextmanager
def group(name, log=print):
    """
    Group outputs in the given with block.
    This does nothing in non GitHub Actions environment for now.
    """
    if in_github_actions:
        log(f"::group::{name}")
    try:
        yield
    finally:
        if in_github_actions:
            log("::endgroup::")
```

Then we can keep `log` integration-specific.

Comment (kou) [reply dev/archery/archery/utils/logger.py]: I've used this style.

**Comment by kou** [comment]
Grouped:

https://github.com/kou/arrow/actions/runs/10822024480/job/30025155262#step:9:21

```text
kernel.core_pattern = /tmp/core.%e.%p
DEBUG:archery:Executing `['git', '-C', '/home/runner/work/arrow/arrow', 'rev-parse', '--show-toplevel']`
Docker: Preppare
Docker: Pull
Docker: Build
DEBUG:archery:Executing `['docker', 'compose', '--file=/home/runner/work/arrow/arrow/docker-compose.yml', 'run', '--rm', '-e', 'ARCHERY_DEFAULT_BRANCH=main', '-e', 'ARCHERY_INTEGRATION_WITH_NANOARROW=1', '-e', 'ARCHERY_INTEGRATION_WITH_RUST=1', 'conda-integration']`
 Network arrow_default  Creating
 Network arrow_default  Created
+ /arrow/ci/scripts/rust_build.sh /arrow /build
Integration: Build: Rust
Integration: Build: nanoarrow
+ '[' ON == ON ']'
+ /arrow/ci/scripts/cpp_build.sh /arrow /build
Integration: Build: C++
+ pip install pythonnet
Integration: Prepare: Dependencies
fatal: not a git repository (or any of the parent directories): .git
Integration: Test: IPC: C++ -> C++
Integration: Test: IPC: C++ -> Java
Integration: Test: IPC: C++ -> JS
Integration: Test: IPC: C++ -> C#
Integration: Test: IPC: C++ -> Go
Integration: Test: IPC: C++ -> Rust
Integration: Test: IPC: Java -> C++
Integration: Test: IPC: Java -> Java
Integration: Test: IPC: Java -> JS
Integration: Test: IPC: Java -> C#
Integration: Test: IPC: Java -> Go
Integration: Test: IPC: Java -> Rust
Integration: Test: IPC: JS -> C++
Integration: Test: IPC: JS -> Java
Integration: Test: IPC: JS -> JS
Integration: Test: IPC: JS -> C#
Integration: Test: IPC: JS -> Go
Integration: Test: IPC: JS -> Rust
Integration: Test: IPC: C# -> C++
Integration: Test: IPC: C# -> Java
Integration: Test: IPC: C# -> JS
Integration: Test: IPC: C# -> C#
Integration: Test: IPC: C# -> Go
Integration: Test: IPC: C# -> Rust
Integration: Test: IPC: Go -> C++
Integration: Test: IPC: Go -> Java
Integration: Test: IPC: Go -> JS
Integration: Test: IPC: Go -> C#
Integration: Test: IPC: Go -> Go
Integration: Test: IPC: Go -> Rust
Integration: Test: IPC: Rust -> C++
Integration: Test: IPC: Rust -> Java
Integration: Test: IPC: Rust -> JS
Integration: Test: IPC: Rust -> C#
Integration: Test: IPC: Rust -> Go
Integration: Test: IPC: Rust -> Rust
Integration: Test: IPC: Gold: C++
Integration: Test: IPC: C++ -> C++

Integration: Test: IPC: Gold: Java
Integration: Test: IPC: Java -> Java

...
```

**Comment by kou** [comment]
I'll merge this.

**Pull request merged**